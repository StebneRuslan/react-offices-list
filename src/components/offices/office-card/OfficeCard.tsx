import React from 'react';
import {
  OfficeAdditionalDetailsContainer,
  OfficeDetails,
  OfficeAddressDetailsContainer,
  OfficeCardWrapper,
  OfficeDetailsText,
  OfficeDetailsTitle,
  OfficeDetailsContainer
} from './elements';
import { Button } from '../../shared/Button';
import { OfficeControlsWrapper } from '../elements';
import { pick } from 'lodash';

export function OfficeCard (props: any) {
  const renderPrimary = (value: boolean) => value ? <OfficeDetailsText>Primary HQ</OfficeDetailsText> : null;
  const renderAddressDetails = () => {
    const additionalAddressDetails = Object.values(pick(props.office, ['city', 'postalCode'])).join(', ');
    const details = Object.values(pick(props.office, ['streetAddress', 'country', 'secondAddress']));
    return [additionalAddressDetails].concat(details).map((element, index) => <OfficeDetailsText key={index}>{element}</OfficeDetailsText>);
  };
  const renderUserDetails = () => {
    const details: any = pick(props.office, ['phone', 'fax', 'email']);
    const result = [];
    for(let key in details) {
      result.push(
        <OfficeDetailsContainer key={key}>
          <OfficeDetailsTitle>{key}</OfficeDetailsTitle>
          <OfficeDetailsText>{details[key]}</OfficeDetailsText>
        </OfficeDetailsContainer>
      )
    }
    return result;
  };
  return (
    <OfficeCardWrapper>
      <OfficeAddressDetailsContainer>
        <OfficeDetailsTitle>Address</OfficeDetailsTitle>
        <OfficeDetails>
          {renderPrimary(props.office.officeType)}
          {renderAddressDetails()}
        </OfficeDetails>
      </OfficeAddressDetailsContainer>

      <OfficeAdditionalDetailsContainer>
        <OfficeDetails>
          {renderUserDetails()}
        </OfficeDetails>
      </OfficeAdditionalDetailsContainer>

      <OfficeControlsWrapper>
        <Button white onClick={() => props.removeOffice(props.office.name)}>Remove</Button>
        <Button blue onClick={() => props.toggleEditableOfficeForm(props.office.name)}>Edit</Button>
      </OfficeControlsWrapper>
    </OfficeCardWrapper>
  );
}
