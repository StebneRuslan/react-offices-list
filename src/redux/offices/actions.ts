import actionTypes from './action-types'
import { OfficeModel } from '../../models/office.model';

const toggleNewOfficeForm = () => ({ type: actionTypes.TOGGLE_NEW_OFFICE_FORM });
const toggleEditableOfficeForm = (data: string) => ({ type: actionTypes.TOGGLE_EDITABLE_OFFICE_FORM, payload: data });

const getOfficeList = () => ({ type: actionTypes.GET_OFFICES });
const getOfficeListSuccess = (data: OfficeModel[]) => ({ type: actionTypes.GET_OFFICES_SUCCESS, payload: data });
const getOfficeListError = (data: string) => ({ type: actionTypes.GET_OFFICES_ERROR, err: data });

const createOffice = (data: OfficeModel) => ({ type: actionTypes.CREATE_OFFICE, payload: data});
const createOfficeSuccess = (data: OfficeModel) => ({ type: actionTypes.CREATE_OFFICE_SUCCESS, payload: data});
const createOfficeError = (data: string) => ({ type: actionTypes.CREATE_OFFICE_ERROR, err: data});

const removeOffice = (data: string) => ({ type: actionTypes.REMOVE_OFFICE, payload: data});
const removeOfficeSuccess = (data: string) => ({ type: actionTypes.REMOVE_OFFICE_SUCCESS, payload: data});
const removeOfficeError = (data: string) => ({ type: actionTypes.REMOVE_OFFICE_ERROR, err: data});

const updateOffice = (data: OfficeModel) => ({ type: actionTypes.UPDATE_OFFICE, payload: data});
const updateOfficeSuccess = (data: OfficeModel) => ({ type: actionTypes.UPDATE_OFFICE_SUCCESS, payload: data});
const updateOfficeError = (data: string) => ({ type: actionTypes.UPDATE_OFFICE_ERROR, payload: data});

export default {
  toggleNewOfficeForm,
  toggleEditableOfficeForm,
  getOfficeList,
  getOfficeListSuccess,
  getOfficeListError,
  createOffice,
  createOfficeSuccess,
  createOfficeError,
  removeOffice,
  removeOfficeSuccess,
  removeOfficeError,
  updateOffice,
  updateOfficeSuccess,
  updateOfficeError
}
