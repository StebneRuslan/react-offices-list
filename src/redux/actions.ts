import offices from './offices/actions'

export default {
  offices
}
