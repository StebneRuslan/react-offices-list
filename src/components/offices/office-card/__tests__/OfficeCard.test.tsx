import React from 'react';
import { shallow, configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { OfficeCard } from '../OfficeCard';
import actions from '../../../../redux/offices/actions';

configure({ adapter: new Adapter() });

let props = {
  office: {
    officeType: false,
    name: 'name'
  },
  removeOffice: jest.fn().mockReturnValue(actions.removeOffice('name')),
  toggleEditableOfficeForm: jest.fn().mockReturnValue(actions.toggleEditableOfficeForm('name'))
};

describe('<OfficeCard />', () => {
  it('OfficeCard renders without crashing', () => {
    const component = shallow(<OfficeCard { ...props } />);
    expect(component).toMatchSnapshot()
  });

  it('OfficeCard renders without crashing', () => {
    const wrapper = shallow(<OfficeCard { ...props } />);
    expect(wrapper.exists()).toBe(true);
  });

  it('OfficeCard renders with props', () => {
    const wrapper = mount(<OfficeCard { ...props } />);
    expect(wrapper.props().office.officeType).toBe(false);
  });

  it('Should call the removeOffice function when clicked', () => {
    const wrapper = mount(<OfficeCard { ...props } />);
    expect(props.removeOffice.mock.calls.length).toBe(0);
    wrapper.find('#remove-office-button').at(0).simulate('click');
    expect(props.removeOffice).toHaveBeenCalledWith('name');
  });

  it('Should call the toggleEditableOfficeForm function when clicked', () => {
    const wrapper = mount(<OfficeCard { ...props } />);
    expect(props.toggleEditableOfficeForm.mock.calls.length).toBe(0);
    wrapper.find('#toggle-editable-office-button').at(0).simulate('click');
    expect(props.toggleEditableOfficeForm).toHaveBeenCalledWith('name');
  });
});
