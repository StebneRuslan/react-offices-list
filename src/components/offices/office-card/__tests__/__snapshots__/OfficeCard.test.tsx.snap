// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OfficeCard /> OfficeCard renders without crashing 1`] = `
<styled.div>
  <styled.div>
    <Styled(styled.span)>
      Address
    </Styled(styled.span)>
    <styled.div>
      <styled.p
        key="0"
      />
    </styled.div>
  </styled.div>
  <Styled(styled.div)>
    <styled.div />
  </Styled(styled.div)>
  <styled.div>
    <styled.button
      id="remove-office-button"
      onClick={[Function]}
      white={true}
    >
      Remove
    </styled.button>
    <styled.button
      blue={true}
      id="toggle-editable-office-button"
      onClick={[Function]}
    >
      Edit
    </styled.button>
  </styled.div>
</styled.div>
`;
